### 4.05 Consistency check
#### 1 Importing libraries
#### 2 Importing dataframes
#### 3 Consistency check of df_prods

### 1 Importing libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

### 2 Importing dataframes

#### -importing df_prods from “Original Data”  

In [2]:
path=r'C:\Users\user\07-2023 Instacart Basket Analysis\02 Data'

In [3]:
path

'C:\\Users\\user\\07-2023 Instacart Basket Analysis\\02 Data'

In [4]:
df_prods=pd.read_csv(os.path.join(path,'Original Data','products.csv' ), index_col = False)

In [5]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

#### - importing df_ords from “Prepared Data”  using the same path

In [6]:
df_ords=pd.read_csv(os.path.join(path,'Prepared Data', 'orders_wrangled.csv'), index_col = 0)

In [7]:
df_ords

order_id  user_id eval_set  order_number  day_of_week  \
0         2539329        1    prior             1            2   
1         2398795        1    prior             2            3   
2          473747        1    prior             3            3   
3         2254736        1    prior             4            4   
4          431534        1    prior             5            4   
...           ...      ...      ...           ...          ...   
3421078   2266710   206209    prior            10            5   
3421079   1854736   206209    prior            11            4   
3421080    626363   206209    prior            12            1   
3421081   2977660   206209    prior            13            1   
3421082    272231   206209    train            14            6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

### 3 Consistency check of df_prods

In [8]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

#### df_prods don't have any missing volums because Count is the same for all columns. The price column has very big Max value, there is a mistaken max value to correct.

### 4.5.3-2 Consistency check of df_ords

In [9]:
df_ords.describe()

order_id       user_id  order_number   day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

#### There are missing values in 'days_since_prior_order' because the value of Count is different from other columns. It seems there are duplicates in order+number because the 50% statistic value is 10 times less then max value.

### 4 checking whether a df_ords contains any mixed-type columns 

In [10]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

#### there are no columns with mixed data type

### 4.5.5-1checking df_ords and df_prods for missing values

In [11]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
day_of_week                    0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### the column 'days_since_prior_order' has 206,209 missing values. It can be because the date of prior order isn't known or these rows are first order data and prior order don't exist

In [12]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

### 4.5.5-2 visualisation of missing values: creating df_nan containing only missing values

#### ---for df_prods

In [13]:
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [15]:
df_nan.shape

(16, 5)

#### ---for ords

In [16]:
df_nan=df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [17]:
df_nan

order_id  user_id eval_set  order_number  day_of_week  \
0         2539329        1    prior             1            2   
11        2168274        2    prior             1            2   
26        1374495        3    prior             1            1   
39        3343014        4    prior             1            6   
45        2717275        5    prior             1            3   
...           ...      ...      ...           ...          ...   
3420930    969311   206205    prior             1            4   
3420934   3189322   206206    prior             1            3   
3421002   2166133   206207    prior             1            6   
3421019   2227043   206208    prior             1            1   
3421069   3154581   206209    prior             1            3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

### 4.5.6 Addressing the missing values using an appropriate method 

#### -in df_ords

In [18]:
df_ords['days_since_prior_order']=df_ords['days_since_prior_order'].fillna(0)

In [19]:
df_ords.head()

order_id  user_id eval_set  order_number  day_of_week  order_hour_of_day  \
0   2539329        1    prior             1            2                  8   
1   2398795        1    prior             2            3                  7   
2    473747        1    prior             3            3                 12   
3   2254736        1    prior             4            4                  7   
4    431534        1    prior             5            4                 15   

   days_since_prior_order  
0                     0.0  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

#### The missing datas emerged because there are no prior orders before the first orders. Therefore, the intire rows anyway make sense and I didn't like to detet them. So the decision is to implement 0 values instead of NaNs

#### -in df_prods

In [20]:
df_prods['product_name'].fillna('noname', inplace=True)

In [21]:
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [22]:
df_nan

Empty DataFrame
Columns: [product_id, product_name, aisle_id, department_id, prices]
Index: []

### 4.5.7 Checking for duplicate values in df_ords 

In [23]:
df_ords_dupl=df_ords[df_ords.duplicated()]

In [24]:
df_ords_dupl

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

### 4.5.7-2 checking for duplicates in df_prods

In [25]:
df_prods_dupl=df_prods[df_prods.duplicated()]

In [26]:
df_prods_dupl

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

### 4.5.8 Addressing the duplicates

In [27]:
df_prods.shape

(49693, 5)

In [28]:
df_prods_no_duplicates=df_prods.drop_duplicates()

In [29]:
df_prods_no_duplicates.shape

(49688, 5)

#### We dont need the duplicates and therefore I have canseled this rows

### 4.5.9 Exporting df

In [30]:
df_prods_no_duplicates.to_csv(os.path.join(path,'Prepared Data', 'prods_no_duplicates.csv'))

In [31]:
df_ords.to_csv(os.path.join(path, 'Prepared Data', 'ords_no_NaNs.csv'))